pkg/controller/externalnode/controller_test.go

```diff
@@ -0,0 +1,652 @@
+				ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+				Spec: v1alpha1.ExternalNodeSpec{
+					Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+				},
+			},
+			expectedEntities: []v1alpha2.ExternalEntity{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "vm1",
+						Namespace: "ns1",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "crd.antrea.io/v1alpha1",
+								Kind:       "ExternalNode",
+								Name:       "vm1",
+							},
+						},
+						Labels: map[string]string{"en": "vm1"},
+					},
+					Spec: v1alpha2.ExternalEntitySpec{
+						Endpoints: []v1alpha2.Endpoint{
+							{IP: "1.1.1.2"},
+						},
+						ExternalNode: "vm1",
+					},
+				},
+			},
+		},
+		{
+			externalNode: &v1alpha1.ExternalNode{
+				ObjectMeta: metav1.ObjectMeta{Name: "vm2", Namespace: "ns1", Labels: map[string]string{"en": "vm2"}},
+				Spec: v1alpha1.ExternalNodeSpec{
+					Interfaces: []v1alpha1.NetworkInterface{{Name: "ens192", IPs: []string{"1.1.1.3"}}},
+				},
+			},
+			expectedEntities: []v1alpha2.ExternalEntity{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      generateEntityName("vm2", "ens192"),
+						Namespace: "ns1",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "crd.antrea.io/v1alpha1",
+								Kind:       "ExternalNode",
+								Name:       "vm2",
+							},
+						},
+						Labels: map[string]string{"en": "vm2"},
+					},
+					Spec: v1alpha2.ExternalEntitySpec{
+						Endpoints: []v1alpha2.Endpoint{
+							{Name: "ens192", IP: "1.1.1.3"},
+						},
+						ExternalNode: "vm2",
+					},
+				},
+			},
+		},
+		{
+			externalNode: &v1alpha1.ExternalNode{
+				ObjectMeta: metav1.ObjectMeta{Name: "vm3", Namespace: "ns1", Labels: map[string]string{"en": "vm3"}},
+				Spec: v1alpha1.ExternalNodeSpec{
+					Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.4", "2.2.2.4"}}},
+				},
+			},
+			expectedEntities: []v1alpha2.ExternalEntity{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "vm3",
+						Namespace: "ns1",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "crd.antrea.io/v1alpha1",
+								Kind:       "ExternalNode",
+								Name:       "vm3",
+							},
+						},
+						Labels: map[string]string{"en": "vm3"},
+					},
+					Spec: v1alpha2.ExternalEntitySpec{
+						Endpoints: []v1alpha2.Endpoint{
+							{IP: "1.1.1.4"},
+							{IP: "2.2.2.4"},
+						},
+						ExternalNode: "vm3",
+					},
+				},
+			},
+		},
+	} {
+		testAddExternalNode(t, tc.externalNode, tc.expectedEntities)
+	}
+}
+
+func TestUpdateExternalNode(t *testing.T) {
+	for _, tc := range []struct {
+		externalNode             *v1alpha1.ExternalNode
+		existingEntity           v1alpha2.ExternalEntity
+		updatedExternalNode      *v1alpha1.ExternalNode
+		expectedEntity           v1alpha2.ExternalEntity
+		entityExistsBeforeUpdate bool
```

Comment:
why it needs `entityExistsBeforeUpdate`? Can't it use nil existingEntity to represent not existing before?